src/_P021_Level.ino

```diff
@@ -60,111 +204,290 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+        uint32_t force_time    = P021_FORCE_TIME;        // Value for forces run
+
+        const __FlashStringHelper *unit1 = F("seconds"); // use minutes or seconds
+        const __FlashStringHelper *unit2 = F("seconds"); // use hours or seconds
+
+        if (bitRead(flags, P021_LONG_TIMER_UNIT))
+        {
+          min_time      = seconds2minutes(P021_MIN_TIME);
+          interval_time = seconds2hours(P021_INTERVAL_TIME);
+          force_time    = seconds2minutes(P021_FORCE_TIME);
+          unit1         = F("minutes");
+          unit2         = F("hours");
+        }
+
+        // Minimum on time
+        addFormNumericBox(F("Minimum running time"), F(P021_GUID_MIN_TIME), min_time, 0);
+        addUnit(unit1);
+
+        // Interval time (max idle time)
+        addFormNumericBox(F("Maximum idle time"), F(P021_GUID_INTERVAL_TIME), interval_time, 0);
+        addUnit(unit2);
+
+        // Interval circulation time (forced circulation)
+        addFormNumericBox(F("Forced circulation time"), F(P021_GUID_FORCE_TIME), force_time, 1);
+        addUnit(unit1);
+        #endif // ifdef P021_MIN_BUILD_SIZE - else
+
+
+        // Symetrical/asymetrical hysteresis [checkbox]
+        addFormCheckBox(F("Symetrical hysteresis"), F(P021_GUID_SYM_HYSTERESIS), bitRead(flags, P021_SYM_HYSTERESIS));
+
+        // Extension period time calculated from start or stop time [checkbox]
+        addFormCheckBox(F("Extend at end"),         F(P021_GUID_EXTEND_END),     bitRead(flags, P021_EXTEND_END));
+
+        // Direction of input comparator ( normal = input>setpoint; invert = input<setpoint) [checkbox]
+        addFormCheckBox(F("Invert Input"),          F(P021_GUID_INV_INPUT),      bitRead(flags, P021_INV_INPUT));
+
+        // Control evaluation at low speed (1Hz) or high speed (10Hz) [checkbox]
+        addFormCheckBox(F("Slow evaluation"),       F(P021_GUID_SLOW_EVAL),      bitRead(flags, P021_SLOW_EVAL));
+
+        // Provide the controller state as second sensor output value [checkbox]
+        addFormCheckBox(F("State as output value"), F(P021_GUID_STATE_OUTP),     bitRead(flags, P021_STATE_OUTP));
+      }
 
       success = true;
       break;
     }
 
     case PLUGIN_WEBFORM_SAVE:
     {
-      P021_CHECK_TASK          = getFormItemInt(F("ptask"));
-      P021_CHECK_VALUE         = getFormItemInt(F("pvalue"));
-      P021_DONT_ALWAYS_SAVE    = isFormItemChecked(F("psave_always")) ? 0 : 1; // inverted flag!
-      P021_TRIGGER_LEVEL       = getFormItemFloat(F("psetvalue"));
-      P021_TRIGGER_LAST_STORED = P021_TRIGGER_LEVEL;
-      P021_TRIGGER_HYSTERESIS  = getFormItemFloat(F("physt"));
-      P021_AUTOSAVE_TIMER      = getFormItemInt(F("pautosave")) * 60; // Store in seconds
-      P021_INVERT_OUTPUT       = isFormItemChecked(F("inv")) ? 1 : 0;
+      const bool newExtFunct  =  (getFormItemInt(F(P021_GUID_EXT_FUNCT)) != 0);
+      uint16_t   flagRegister = P021_FLAGS; // Reduce expensive access to P021 to write new flags
```

Comment:
This variable can also be a `const`, and for consistency should be renamed to `flags`. The other `flags` variable is not in this scope, so no conflicts.

Reply:
This is not a const as it is updated by successive bitwrite() calls. This is where most of the flash memory is saved. Indeed, forgot to rename it when reviewing it myself.